### PR TITLE
WIP: #28 Add containsKey assertion

### PR DESCRIPTION
**Pull request by uaArsen**
[#28](https://github.com/robstoll/atrium/issues/28)

Not sure that I done it wright, especially testing part with Spec is confusing me.

### CODE REVIEW

**Comment by robstoll** [comment]
Thank you very much for your contribution. Looks quite good to me, I'll change a few things and fix the spec unless you would like to give it a try.

On the other hand, I need a confirmation for the Contributor Agreements from you. Please edit your description or add a comment stating:

I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

**Comment by uaArsen** [comment]
Hi, please give me some hints. I would try to make changes.

File: apis/cc-de_CH/atrium-api-cc-de_CH-common/src/main/kotlin/ch/tutteli/atrium/api/cc/de_CH/mapAssertions.kt [hunk 1]
```diff
@@ -6,6 +6,15 @@ import ch.tutteli.atrium.creating.AssertionPlant
 import ch.tutteli.atrium.creating.AssertionPlantNullable
 import ch.tutteli.atrium.domain.builders.AssertImpl
 
+/**
+ * Makes the assertion that [AssertionPlant.subject]'s [Map] contains [key].
+ *
+ * @return This plant to support a fluent API.
+ * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
+ */
+fun <K, V> Assert<Map<K, V>>.enthaeltKey(key: K)
```
Comment (robstoll) [inline apis/cc-de_CH/atrium-api-cc-de_CH-common/src/main/kotlin/ch/tutteli/atrium/api/cc/de_CH/mapAssertions.kt]: Remove type parameter V as you don't use it within the function, use `Assert<Map<K, *>>` instead

File: apis/cc-en_GB/atrium-api-cc-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/en_GB/mapAssertions.kt [hunk 2]
```diff
@@ -6,6 +6,16 @@ import ch.tutteli.atrium.creating.AssertionPlant
...
 import ch.tutteli.atrium.domain.builders.AssertImpl
 
+
+/**
+ * Makes the assertion that [AssertionPlant.subject]'s [Map] contains [key].
+ *
+ * @return This plant to support a fluent API.
+ * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
+ */
+fun <K, V> Assert<Map<K, V>>.containsKey(key: K)
```
Comment (robstoll) [inline apis/cc-en_GB/atrium-api-cc-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/en_GB/mapAssertions.kt]: remove `V`

File: apis/cc-de_CH/atrium-api-cc-de_CH-common/src/main/kotlin/ch/tutteli/atrium/api/cc/de_CH/mapAssertions.kt [hunk 3]
```diff
@@ -6,6 +6,15 @@ import ch.tutteli.atrium.creating.AssertionPlant
 import ch.tutteli.atrium.creating.AssertionPlantNullable
 import ch.tutteli.atrium.domain.builders.AssertImpl
 
+/**
+ * Makes the assertion that [AssertionPlant.subject]'s [Map] contains [key].
```
Comment (robstoll) [inline apis/cc-de_CH/atrium-api-cc-de_CH-common/src/main/kotlin/ch/tutteli/atrium/api/cc/de_CH/mapAssertions.kt]: I would rephrase to 
`Makes the assertion that [AssertionPlant.subject] contains the given [key].`

File: apis/cc-en_GB/atrium-api-cc-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/en_GB/mapAssertions.kt [hunk 4]
```diff
@@ -6,6 +6,16 @@ import ch.tutteli.atrium.creating.AssertionPlant
 import ch.tutteli.atrium.creating.AssertionPlantNullable
 import ch.tutteli.atrium.domain.builders.AssertImpl
 
+
+/**
+ * Makes the assertion that [AssertionPlant.subject]'s [Map] contains [key].
```
Comment (robstoll) [inline apis/cc-en_GB/atrium-api-cc-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/en_GB/mapAssertions.kt]: rephrase

File: apis/cc-infix-en_GB/atrium-api-cc-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/infix/en_GB/mapAssertions.kt [hunk 5]
```diff
@@ -8,6 +8,15 @@ import ch.tutteli.atrium.creating.Assert
 import ch.tutteli.atrium.creating.AssertionPlant
 import ch.tutteli.atrium.domain.builders.AssertImpl
 
+/**
+ * Makes the assertion that [AssertionPlant.subject]'s [Map] contains [key].
```
Comment (robstoll) [inline apis/cc-infix-en_GB/atrium-api-cc-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/infix/en_GB/mapAssertions.kt]: rephrase

File: apis/cc-infix-en_GB/atrium-api-cc-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/infix/en_GB/mapAssertions.kt [hunk 6]
```diff
@@ -8,6 +8,15 @@ import ch.tutteli.atrium.creating.Assert
...
+ *
+ * @return This plant to support a fluent API.
+ * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
+ */
+infix fun <K, V> Assert<Map<K, V>>.containsKey(key: K)
```
Comment (robstoll) [inline apis/cc-infix-en_GB/atrium-api-cc-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/infix/en_GB/mapAssertions.kt]: remove `V`

File: domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/mapAssertions.kt [hunk 7]
```diff
@@ -13,6 +13,11 @@ import ch.tutteli.atrium.translations.DescriptionBasic
 import ch.tutteli.atrium.translations.DescriptionCollectionAssertion
 import ch.tutteli.atrium.translations.DescriptionMapAssertion
 
+fun <K, V> _containsKey(plant: AssertionPlant<Map<K, V>>, key: K): Assertion = AssertImpl.builder.descriptive
+    .withTest { plant.subject.containsKey(key) }
+    .withDescriptionAndRepresentation(DescriptionMapAssertion.MAP_CONTAINS, key.toString())
```
Comment (robstoll) [inline domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/mapAssertions.kt]: please don't use `toString()` otherwise error reporting is misleading. For instance: 
``` 
assert(mapOf(1 to "hello", 3 to "world")).containsKey(2)
```
would report:
```
assert: {1="hello", 3="world"}
- contains key: "2" 
```
Which makes it look like one inadvertently passed `"2"` instead of `2`

File: misc/atrium-spec/src/main/kotlin/ch/tutteli/atrium/spec/integration/MapAssertionsSpec.kt [hunk 8]
```diff
@@ -41,13 +45,27 @@ abstract class MapAssertionsSpec(
...
     val (isEmpty, isEmptyFun) = isEmptyPair
     val (isNotEmpty, isNotEmptyFun) = isNotEmptyPair
 
     val isDescr = DescriptionBasic.IS.getDefault()
     val isNotDescr = DescriptionBasic.IS_NOT.getDefault()
     val empty = DescriptionCollectionAssertion.EMPTY.getDefault()
+    val containsKeyDescr = DescriptionMapAssertion.MAP_CONTAINS.getDefault()
+
+    describeFun(containsKey) {
+        it("does not throw if a map contains key a") {
```
Comment (robstoll) [inline misc/atrium-spec/src/main/kotlin/ch/tutteli/atrium/spec/integration/MapAssertionsSpec.kt]: "does not throw if the map contains the key"

File: misc/atrium-spec/src/main/kotlin/ch/tutteli/atrium/spec/integration/MapAssertionsSpec.kt [hunk 9]
```diff
@@ -41,13 +45,27 @@ abstract class MapAssertionsSpec(
...
+            assert(mapOf()).containsKeyFun("a")
+        }
+
+        it("throws an AssertionError if map not contains key") {
```
Comment (robstoll) [inline misc/atrium-spec/src/main/kotlin/ch/tutteli/atrium/spec/integration/MapAssertionsSpec.kt]: throws an AssertionError if the map does not contain the key

File: translations/en_GB/atrium-translations-en_GB-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionMapAssertion.kt [hunk 10]
```diff
@@ -8,5 +8,6 @@ import ch.tutteli.atrium.reporting.translating.StringBasedTranslatable
  */
 enum class DescriptionMapAssertion(override val value: String) : StringBasedTranslatable {
     CANNOT_EVALUATE_KEY_DOES_NOT_EXIST("$COULD_NOT_EVALUATE_DEFINED_ASSERTIONS -- given key does not exist.\n$VISIT_COULD_NOT_EVALUATE_ASSERTIONS"),
-    KEY_DOES_NOT_EXIST("❗❗ key does not exist")
+    KEY_DOES_NOT_EXIST("❗❗ key does not exist"),
+    MAP_CONTAINS("Map contains key")
```
Comment (robstoll) [inline translations/en_GB/atrium-translations-en_GB-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionMapAssertion.kt]: No need for the MAP suffix as we only have the description for MapAssertions in this enum anyway. Please rephrase it as follows:
`CONTAINS_KEY("contains key")`
Error reporting will then look as shown in an above comment.

Comment (robstoll) [inline translations/en_GB/atrium-translations-en_GB-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionMapAssertion.kt]: Do not forget to add your translation to de_CH => search for second DescriptionMapAssertion. There it should be:
`CONTAINS_KEY("enthält Key")

File: misc/atrium-spec/src/main/kotlin/ch/tutteli/atrium/spec/integration/MapAssertionsSpec.kt [hunk 11]
```diff
@@ -41,13 +45,27 @@ abstract class MapAssertionsSpec(
...
+            expect {
+                assert(mapOf( "b" to 2)).containsKeyFun("a")
+            }.toThrow<AssertionError> { messageContains("$isDescr: $containsKeyDescr")}
```
Comment (robstoll) [inline misc/atrium-spec/src/main/kotlin/ch/tutteli/atrium/spec/integration/MapAssertionsSpec.kt]: We want to check two things here:
- does the message contain the assertion (could fail for another reason)
- does the assertion include your description (`$containsKeyDescr...`) as well as the representation of the key
additional hint, no need for `$isDescr` here

**Comment by robstoll** [comment]
@uaArsen btw. GitHub was not able to identify the commits as your contribution (see for instance https://github.com/uaArsen/atrium/commit/9fbf97844f6126a1485894cca7565793cd2ccb56, the user icon is not your icon). I guess you have another email here: https://github.com/settings/emails and need to add arsenzhd@gmail.com in addition
